nova_vm/src/ecmascript/builtins/indexed_collections/typed_array_objects/abstract_operations.rs

```diff
@@ -0,0 +1,686 @@
+    agent: &mut Agent,
+    mut gc: GcScope<'_, '_>,
+    o: TypedArray,
+    src_array: TypedArray,
+) -> JsResult<()> {
+    let src_heap_data = &agent[src_array];
+
+    // 1. Let srcData be srcArray.[[ViewedArrayBuffer]].
+    let src_data = src_heap_data.viewed_array_buffer;
+
+    // 2. Let elementType be TypedArrayElementType(O).
+    // 3. Let elementSize be TypedArrayElementSize(O).
+    let element_size = size_of::<T>();
+
+    // 4. Let srcType be TypedArrayElementType(srcArray).
+    // 5. Let srcElementSize be TypedArrayElementSize(srcArray).
+    let src_element_size = size_of::<U>();
+
+    // 6. Let srcByteOffset be srcArray.[[ByteOffset]].
+    let src_byte_offset = src_array.byte_offset(agent);
+
+    // 7. Let srcRecord be MakeTypedArrayWithBufferWitnessRecord(srcArray, seq-cst).
+    let src_record =
+        make_typed_array_with_buffer_witness_record(agent, src_array, Ordering::SeqCst);
+
+    // 8. If IsTypedArrayOutOfBounds(srcRecord) is true, throw a TypeError exception.
+    if is_typed_array_out_of_bounds::<U>(agent, &src_record) {
+        return Err(agent.throw_exception_with_static_message(
+            ExceptionType::TypeError,
+            "TypedArray out of bounds",
+        ));
+    }
+
+    // 9. Let elementLength be TypedArrayLength(srcRecord).
+    let element_length = typed_array_length::<T>(agent, &src_record);
+
+    // 10. Let byteLength be elementSize × elementLength.
+    let byte_length = element_size * element_length;
+
+    // 11. If elementType is srcType, then
+    let data = if T::PROTO == U::PROTO {
+        // a. Let data be ? CloneArrayBuffer(srcData, srcByteOffset, byteLength).
+        clone_array_buffer(agent, src_data, src_byte_offset, byte_length)?
+    } else {
+        // 12. Else,
+        // a. Let data be ? AllocateArrayBuffer(%ArrayBuffer%, byteLength).
+        let array_buffer_constructor = agent.current_realm().intrinsics().array_buffer();
+        let data = allocate_array_buffer(
+            agent,
+            array_buffer_constructor.into_function(),
+            byte_length as u64,
+            None,
+        )?;
+
+        // b. If srcArray.[[ContentType]] is not O.[[ContentType]], throw a TypeError exception.
+        if T::IS_BIGINT != U::IS_BIGINT {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "TypedArray content type mismatch",
+            ));
+        }
+
+        // c. Let srcByteIndex be srcByteOffset.
+        let mut src_byte_index = src_byte_offset;
+        // d. Let targetByteIndex be 0.
+        let mut target_byte_index = 0;
+        // e. Let count be elementLength.
+        let mut count = element_length;
+        // f. Repeat, while count > 0,
+        while count != 0 {
+            // i. Let value be GetValueFromBuffer(srcData, srcByteIndex, srcType, true, unordered).
+            let value = get_value_from_buffer::<T>(
```

Comment:
**issue:** Either the get or set side should be `::<U>`, I think? Not sure which since I'm not sure if T or U is "equal" to `srcType`.

Also: Here we really kind of see how problematic it is to use `Value` as the intermediate format. We'd _really_ want to just get `u8` or `u32` or something out of here and not have to really interact with the Agent at all basically... I wonder if we could provide some sort of `as_slice::<T>` and `as_slice_mut::<T>` methods which we could then work with...

Reply:
Yep. I see that it could really do with a smaller refactor to get rid of the juggling between value and native types here. I'll leave it for a future tidying up.